File: android/src/main/java/com/aptoslabs/arculus/csdk/WalletSeedCreateWallet.kt
package com.aptoslabs.arculus.csdk

import com.sun.jna.Pointer

class WalletSeedCreateWallet(wallet: Pointer, private val nbrOfWords: Int) :
  CSDKAPICall<String>(wallet) {
  override suspend fun request(): Array<ByteArray> {
    val len = SizeTByReference()

    val pointer = CSDK.WalletSeedCreateWalletRequest(wallet, len, nbrOfWords)

    if (pointer == Pointer.NULL) throw CSDKAPICallRequestCreationException()

    return arrayOf(pointer.getByteArray(0, len.toIntChecked()))
  }

  override suspend fun response(bytes: ByteArray): String {
    val mnemonicSentenceLength = SizeTByReference()

    val pointer = CSDK.WalletCreateWalletResponse(wallet, bytes, bytes.size, mnemonicSentenceLength)

    val data = pointer.getByteArray(0, mnemonicSentenceLength.toIntChecked())

    return data.toString(Charsets.UTF_8).trimEnd('\u0000')
  }
}

File: ios/Arculus/CSDK/WalletSeedCreateWallet.swift
import CSDK

final class WalletSeedCreateWallet: CSDKAPICall<String> {
    private let nbrOfWords: Int
    
    init(wallet: OpaquePointer, nbrOfWords: Int) {
        self.nbrOfWords = nbrOfWords
        
        super.init(wallet: wallet)
    }
    
    override func request() async throws -> [Data] {
        var len: size_t = 0
        
        guard let bytes = WalletSeedCreateWalletRequest(wallet, &len, nbrOfWords) else {
            throw CSDKAPICallError(.requestCreationFailed("seed create wallet"))
        }
        
        return [Data(bytes: bytes, count: len)]
    }
    
    override func response(bytes: [UInt8]) async throws -> ResponseType {
        var mnemonicSentenceLength: size_t = 0
        
        guard let response = WalletSeedCreateWalletResponse(wallet, bytes, bytes.count, &mnemonicSentenceLength) else {
            throw CSDKAPICallError(.responseParsingFailed("seed create wallet"))
        }
        
        let data = Data(bytes: response, count: mnemonicSentenceLength)
        
        guard let phrase = String(data: data, encoding: .utf8) else {
            throw CSDKAPICallError(.responseParsingFailed("seed create wallet"))
        }
        
        return phrase.trimmingCharacters(in: .controlCharacters)
    }
}
